ingeniamotion/fsoe.py

```diff
@@ -0,0 +1,264 @@
+        RPDOMapItem(size_bits=STO_COMMAND_PADDING_SIZE_BITS),
+        RPDOMapItem(size_bits=CRC_O_SIZE_BITS),
+        RPDOMapItem(size_bits=CONN_ID_SIZE_BITS),
+    ]
+    SAFETY_SLAVE_PDU_PDO_ITEMS = [
+        TPDOMapItem(size_bits=FSOE_COMMAND_SIZE_BITS),
+        TPDOMapItem(size_bits=STO_COMMAND_SIZE_BITS),
+        TPDOMapItem(size_bits=STO_COMMAND_PADDING_SIZE_BITS),
+        TPDOMapItem(size_bits=CRC_O_SIZE_BITS),
+        TPDOMapItem(size_bits=CONN_ID_SIZE_BITS),
+    ]
+
+    @classmethod
+    def configure_rpdo_map(cls, rpdo_map: RPDOMap) -> None:
+        """Configure the RPDOMap used for the Safety Master PDU.
+
+        Args:
+            rpdo_map: The RPDOMap instance.
+
+        """
+        rpdo_map.map_register_index = cls.FSOE_RPDO_MAP_1_INDEX
+        for rpdo_map_item in cls.SAFETY_MASTER_PDU_PDO_ITEMS:
+            rpdo_map.add_item(rpdo_map_item)
+
+    @classmethod
+    def configure_tpdo_map(cls, tpdo_map: TPDOMap) -> None:
+        """Configure the TPDOMap used for the Safety Slave PDU.
+
+        Args:
+            tpdo_map: The TPDOMap instance.
+
+        """
+        tpdo_map.map_register_index = cls.FSOE_TPDO_MAP_1_INDEX
+        for tpdo_map_item in cls.SAFETY_SLAVE_PDU_PDO_ITEMS:
+            tpdo_map.add_item(tpdo_map_item)
+
+
+class FSoEMaster:
+    """FSoE Master.
+
+    Args:
+        motion_controller: The MotionController instance.
+
+    """
+
+    DEFAULT_WATCHDOG_TIMEOUT_S = 1
+    DEFAULT_FSOE_SLAVE_ADDRESS = 0
+
+    def __init__(self, motion_controller: "MotionController") -> None:
+        self.logger = ingenialogger.get_logger(__name__)
+        self.__mc = motion_controller
+        self.__handlers: Dict[str, FSoEMasterHandler] = {}
+        self.__latest_connection_id = 1
+
+    def create_fsoe_master_handler(
+        self,
+        servo: str = DEFAULT_SERVO,
+        fsoe_master_watchdog_timeout: float = DEFAULT_WATCHDOG_TIMEOUT_S,
+    ) -> None:
+        """Create an FSoE Master handler linked to a Safe servo drive.
+
+        Args:
+            servo: servo alias to reference it. ``default`` by default.
+            fsoe_master_watchdog_timeout: The FSoE master watchdog timeout in seconds.
+
+        """
+        # TODO: use function to read the FSoE slave address (INGM-446)
+        slave_address = self.DEFAULT_FSOE_SLAVE_ADDRESS
+        master_handler = FSoEMasterHandler(
+            slave_address, self.__latest_connection_id, fsoe_master_watchdog_timeout
+        )
+        self.__handlers[servo] = master_handler
+        self.__latest_connection_id += 1
+
+    def start_master(self, start_pdos: bool = False) -> None:
+        """Start all the FSoE Master handlers.
+
+        Args:
+            start_pdos: if ``True``, start the PDO exchange, if ``False``
+                the PDO exchange should be started after. ``False`` by default.
+
+        """
+        for servo, master_handler in self.__handlers.items():
+            master_handler.start()
+        if start_pdos:
```

Comment:
Even though the fsoe master should reconnect after a few reset cycles typical workflow is to start pdos before the fsoe